src/background/api.ts

```diff
@@ -1,93 +1,78 @@
-  target.feedbackState.images.forEach(image => tweetData.append('images', image.blob, image.name))
-
-  return tweetData
-}
-
-function makeTweetRequest(formData: FormData): Promise<Response> {
-  return fetch(`${window.roarServerUrl}/v1/feedback`, {
-    method: 'POST',
-    credentials: 'include',
-    body: formData,
-  })
-}
-
-export const postTweet = async (target: FeedbackTarget, chrome: typeof global.chrome, dispatchBackgroundActions: Dispatchers<BackgroundAction>) => {
-  const targetId: FeedbackTargetId = target.id
+export type FetchRoarResult<T> =
+  | { ok: true; data: T }
+  | { ok: false; reason: 'response not json'; text: string; details: string }
+  | { ok: false; reason: 'response not expected data type'; text: string; details: string }
+  | { ok: false; reason: 'bad request'; details: string }
+  | { ok: false; reason: 'unauthorized'; details: string }
+  | { ok: false; reason: 'server down'; details: string }
+  | { ok: false; reason: 'unknown status'; details: string; status: number }
+  | { ok: false; reason: 'network down'; details: string }
 
+export async function fetchRoar<T extends object>(path: string, init: RequestInit, decoder: JsonDecoder.Decoder<T>): Promise<FetchRoarResult<T>> {
   try {
-    dispatchBackgroundActions.postTweetStart({ targetId })
-    const res = await makeTweetRequest(tweetFormData(target))
-    if (res.status !== 201) {
-      return dispatchBackgroundActions.postTweetFailure({
-        targetId,
-        error: { message: await res.text() },
-      })
+    const response = await fetch(`${window.roarServerUrl}/${path}`, { credentials: 'include', ...init })
+
+    if (response.status >= 200 && response.status < 300) {
+      const text = await response.text()
+      try {
+        const json = JSON.parse(text)
+        const decoded = decoder.decode(json)
+        if (decoded.isOk()) {
+          return { ok: true, data: decoded.value }
+        } else {
+          return { ok: false, reason: 'response not expected data type', text, details: decoded.error }
+        }
+      } catch (error) {
+        return { ok: false, reason: 'response not json', details: error.message, text }
+      }
+    }
+    if (response.status === 400) {
+      const details = await response.text()
+      return { ok: false, reason: 'bad request', details }
     }
-    const tweetResult = await res.json()
-    if (!tweetResult.url) {
-      return dispatchBackgroundActions.postTweetFailure({
-        targetId,
-        error: { message: 'Response must include a url' },
-      })
+    if (response.status === 401) {
+      return { ok: false, reason: 'unauthorized', details: '401 received' }
     }
-    chrome.tabs.create({ url: tweetResult.url, active: true })
-    return dispatchBackgroundActions.postTweetSuccess({ targetId })
+    if (response.status >= 500) {
+      const details = await response.text()
+      return { ok: false, reason: 'server down', details }
```

Comment:
What if it's a third-party service is down and we get a 503? For example, if Twitter is down. It would be nice letting the user know that we can't continue because Twitter is down and not because our server is down.

Reply:
Added a test for this.